Rewrite Writing Build Scripts in a consistent docs voice (#22276)
gradle/gradle: subprojects/docs/src/docs/userguide/authoring-builds/writing_build_scripts.adoc

[[writing_build_scripts]]
= Writing Build Scripts

Use a build script to configure a _project_. Each Gradle project corresponds to a software component
that needs to be built, like a library or an application. Each build script is associated
with an object of type link:{groovyDslPath}/org.gradle.api.Project.html[Project].
As the build script executes, it configures this `Project`.

[IMPORTANT]
.Build scripts, Settings scripts, and Init scripts
====
_Build scripts_ define `Project` objects.
Gradle uses other scripts as well:

- <<build_lifecycle.adoc#sec:settings_file,Settings scripts>>
  define link:{groovyDslPath}/org.gradle.api.initialization.Settings.html[Settings] objects.
- <<init_scripts.adoc#init_scripts,Init scripts>>
  define link:{groovyDslPath}/org.gradle.api.invocation.Gradle.html[Gradle] objects.
====

== Properties

> Almost all top-level properties and blocks in a build script are part of the `Project` API.

Reviewer: I agree with this.  Many top-level blocks are extensions added by plugins.

> Almost all top-level properties and blocks in a build script are part of the `Project` API.

Author: ```suggestion
Many top-level properties and blocks in a build script are part of the `Project` API.
```

====

[[groovy-dsl-basics]]
== Helpful Groovy Syntax

> Groovy provides many features to help create DSLs. These features can help you write better build scripts, plugins, and tasks.

Reviewer 2: Perhaps "more idiomatic" rather than "better" - assuming that this is what we mean by "better" here.

Author: ```suggestion
Groovy provides many features to help create DSLs. These features can help you write more idiomatic build scripts, plugins, and tasks.
```

> Groovy provides many features to help create DSLs. These features can help you write more idiomatic build scripts, plugins, and tasks.

Author: ```suggestion
The https://docs.groovy-lang.org/latest/html/documentation/index.html[Groovy language] provides many features to help create DSLs. These features can help you write more idiomatic build scripts, plugins, and tasks.
```

> [TIP]

Reviewer: Why'd we drop the link to Groovy language docs but keep the Kotlin one?

Author: Added back, thank you!

Looking for helpful Kotlin syntax? Check out the link:https://kotlinlang.org/docs/reference/[Kotlin reference documentation] and link:https://kotlinlang.org/docs/tutorials/koans.html[Kotlin Koans].

[[sec:groovy_jdk]]
=== Convenience Methods

Groovy adds convenience methods to the standard Java classes.

> For example, `Iterable` gets an `each` method, which iterates over the elements of the `Iterable`:

Reviewer: I think this is a bad example of an extension - it's too close to the API and functionality already available in the JDK.  Maybe something like `File.text`, which hides a lot of the difficult details of file IO that Java devs should be very familiar with.

Reviewer 2: The thing about something like File.text is that it's really File.setText() and getText(). We shouldn't introduce two Groovy features things at once here. I think each is a good example.

Reviewer: That's true.  I just think that with `.forEach()` in JDK 8, and some form of callback-based iteration basically expected as a language feature anywhere today, `.each()` is not such a good example.  Maybe `List.head()` and `List.tail()`?

Author: It sounds like there are mixed pros and cons for changing this example. Would it be OK to put this code block update on the backburner for now, since the current example is workable? I can revise it later.

=== Last Closure Parameter Syntax

The Gradle DSL uses closures in many places.
You can find out more about closures https://docs.groovy-lang.org/latest/html/documentation/index.html#_closures[here].
When the last parameter of a method is a closure, you can place the closure after the method call:

.Closure as method parameter
====
include::sample[dir="snippets/tutorial/groovy/groovy",files="build.gradle[tags=closureAsLastParam]"]
====

[[sec:closure_delegate]]
=== Closure Delegates

Each closure has a `delegate` object. Groovy uses this delegate to look up variable and method
references to nonlocal variables and closure parameters. Gradle uses this for _configuration closures_,
where the `delegate` object refers to the object being configured.

.Closure delegates
====
include::sample[dir="snippets/tutorial/groovy/groovy",files="build.gradle[tags=closureDelegates]"]
====

[[script-default-imports]]
== Default Imports

To make build scripts more concise, Gradle automatically adds a set of import statements to scripts.
As a result, instead of writing `throw new org.gradle.api.tasks.StopExecutionException()`,
you can write `throw new StopExecutionException()` instead.

> The following list shows the imports Gradle adds to each script:

Reviewer 2: I preferred the previous phrasing, although I can see why this might be better for some readers 👍

Author: ```suggestion
Gradle adds the following imports to each script:
```
How about this?

Reviewer 2: Looks good to me. If you agree, maybe we could add 'implicitly' to emphasise the point, as in 'Gradle implicitly adds ...'

.Gradle default imports
[source,groovy]
----
include::default-imports.txt[]
----

[[sec:groovy_script_variables]]
== Groovy Script Variables

Groovy has two types of script variables:

- Variables declared with a type modifier or `def` are visible within closures but not visible within methods.

> - Variables declared without a modifier are visible anywhere in the script.

Reviewer 2: Can we also make a recommendation here? My personal one would be: Avoid declaring variables without "def". If you need to share an object across scope boundaries,  model it as an extension, task input or whatever else.

Author: ```suggestion
- Variables declared without a modifier are visible anywhere in the script.

Prefer declaring variables with `def`.
If you need to share an object across scope boundaries, use an extension or task input.
```
Does this sound good to you?

Reviewer 2: I feel like my thought was not really complete 😅 I think maybe my bias against omitting 'def' is because if you omit 'def', you are relying on the fact that Gradle scripts are executed with an implicit `this`, which may not be obvious to your teammates who aren't familiar with Groovy/Gradle.

Perhaps: "Prefer declaring variables with `def`, to make their scope explicit."
